ecommerce/cybersource_api.py

```diff
@@ -0,0 +1,99 @@
+from main.settings import ECOMMERCE_DEFAULT_PAYMENT_GATEWAY
+from ecommerce.constants import TRANSACTION_TYPE_REFUND, REFUND_SUCCESS_STATES
+from mitol.payment_gateway.api import (
+    Refund as GatewayRefund,
+    PaymentGateway,
+)
+from mitol.payment_gateway.exceptions import RefundDuplicateException
+
+log = logging.getLogger(__name__)
+
+
+def refund_order(*, order_id: int, **kwargs):
+    """
+    A function that performs refund for a given order id
+
+    Args:
+       order_id (int): Id of the order which is being refunded
+       kwargs (dict): Dictionary of the other attributes that are passed e.g. refund amount, refund reason, unenroll
+       If no refund_amount is provided it will use refund amount from Transaction obj
+       unenroll will never be performed if the refund fails
+
+    Returns:
+        tuple(bool, bool):
+        First bool identifies if the refund operation was a success or failure
+        Second bool identifies if unrenroll was success or failure
+    """
+    refund_amount = kwargs.get("refund_amount")
+    refund_reason = kwargs.get("refund_reason", "")
+    unenroll = kwargs.get("unenroll", False)
+
+    with transaction.atomic():
+        # Get order details for the given order_id (Optimally, This should be a fulfilled order)
+        order = FulfilledOrder.objects.select_for_update().get(id=order_id)
+
+        if order.state != Order.STATE.FULFILLED:
+            log.debug(f"Order with order_id {order_id} is not in fulfilled state.")
+            return False, False
+
+        # Extract transaction_id from order CyberSource json
+        order_recent_transaction = Transaction.objects.filter(order=order_id).first()
+
+        if not order_recent_transaction:
+            log.error(
+                f"There is no associated transaction against order_id {order_id}."
+            )
+            return False, False
+
+        transaction_dict = json.loads(json.dumps(order_recent_transaction.data))
+
+        # This should break if there is no transaction_id in transaction_data because that's needed for refund API
+        transaction_id = transaction_dict["transaction_id"]
+
+        order_amount = refund_amount or transaction_dict.get("req_amount")
+        order_currency = transaction_dict.get("req_currency")
+
+        refund_gateway_request = GatewayRefund(
+            transaction_id=transaction_id,
+            refund_amount=order_amount,
+            refund_currency=order_currency,
+        )
+
+        try:
+            response = PaymentGateway.start_refund(
+                ECOMMERCE_DEFAULT_PAYMENT_GATEWAY,
+                refund_gateway_request,
+            )
+            if response.state in REFUND_SUCCESS_STATES:
+                # There is no default reason for response of a Refund, so we append the response json with the reason
+                # provided by the caller
+                if refund_reason:
+                    response.response_data["reason"] = refund_reason
```

Comment:
Instead of putting this in the response data (we shouldn't be modifying it), can you add a new column to `Transaction` for `reason`? It can default to blank because we won't have a reason for purchases. `Order.refund()` will need to be updated to remove this kind of logic because it's doing a similar thing if `api_response_data` is `None`.